### PR TITLE
Keywords 1 & 2

### PR DESCRIPTION
**Pull request by trisyoungs**
This PR addresses the handling of keywords, greatly improving data locality in `Module`s and `Node`s and converting keyword classes into "managing wrappers".

To summarise the major changes:
1. No data is stored on the keyword - instead, references to member variables are stored instead. This obfuscates the need for `KeywordData<T>`, which has been removed.
2. Keywords are created via a templated `add()` function in `KeywordList`, passing the member variable reference and any other relevant parameters to the keyword constructor via a parameter pack.
3. Setting and getting of keyword data is now greatly reduced as the data are local to where they are most often needed, but templated setter/getter functions have been implemented with a minimal set of classes in order to provide this functionality.

A follow-up PR will remove the `KeywordBase::KeywordDataType` enum and work on simplifying keyword creation and usage in the GUI.

Note: This PR includes the branch `new-keywords-1` (see epic tracking this work #827 for more details on the content).

### CODE REVIEW

File: src/modules/skeleton/init.cpp [hunk 1]
```diff
@@ -7,5 +7,5 @@
 // Perform any necessary initialisation for the Module
 void SkeletonModule::initialise()
 {
-    // 	keywords_.add(new BoolKeyword(true), "ExampleKeyword", "Example keyword description", "<args>");
+    // 	keywords_.add<BoolKeyword>("ExampleKeyword", "Example keyword description", localVariable_);
```
Comment (rprospero) [inline src/modules/skeleton/init.cpp:10]: Can we just make this an empty function?

File: src/modules/test/init.cpp [hunk 2]
```diff
@@ -7,5 +7,5 @@
 // Perform any necessary initialisation for the Module
 void TestModule::initialise()
 {
-    // 	keywords_.add(new BoolKeyword(true), "ExampleKeyword", "Example keyword description", "<args>");
+    // 	keywords_.add<BoolKeyword>("ExampleKeyword", "Example keyword description", localVariable_);
```
Comment (rprospero) [inline src/modules/test/init.cpp:10]: Empty?

**Review by rprospero — APPROVED** [review]
This feels a lot cleaner than the old system.  There's two very minor suggestions.  Ignore my comment about using a template base class.  I realised later on that the trivial implementation would cause problems with storing the pointers.  There's still a way we can do it, but it's outside the scope of this PR.

File: src/modules/skeleton/init.cpp (hunk 1 above)
Comment (trisyoungs) [reply src/modules/skeleton/init.cpp]: Yes we can. I have earmarked tidying of the module directories in a future issue - there are too many files in there with too little content than is sensible.

**Pull request merged**